### PR TITLE
feat: Reset the state on shutting down the flagd resolver (#410)

### PR DESCRIPTION
**Pull request by mehtasankets**
### Notes
I've reset the state in corresponding resolvers instead of adding it in the provider. That way, resolvers can decide when and which state to reset to. Let me know if we should move this logic to provider instead.

### CODE REVIEW

File: providers/flagd/src/test/java/dev/openfeature/contrib/providers/flagd/FlagdProviderTest.java [hunk 1]
```diff
@@ -863,4 +912,29 @@ private FlagdProvider createProvider(GrpcConnector grpc, Cache cache, Supplier<P
...
 
+    // Create an in process provider
+    private FlagdProvider createInProcessProvider() {
+
+        final FlagdOptions flagdOptions = FlagdOptions.builder()
+                .resolverType(Config.Resolver.IN_PROCESS)
+                .deadline(1000)
+                .build();
+        final FlagdProvider provider = new FlagdProvider(flagdOptions);
+        final MockStorage mockStorage = new MockStorage(new HashMap<String, FeatureFlag>(), new LinkedBlockingQueue<StorageState>(List.of(StorageState.OK)));
```
Comment (Kavindu-Dodan) [inline providers/flagd/src/test/java/dev/openfeature/contrib/providers/flagd/FlagdProviderTest.java]: `List.of` won't work as we still maintain support for Java 8

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/FlagdProvider.java [hunk 2]
```diff
@@ -14,6 +14,7 @@
 import dev.openfeature.sdk.Value;
 import lombok.extern.slf4j.Slf4j;
 
+import java.util.List;
```
Comment (Kavindu-Dodan) [inline providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/FlagdProvider.java]: could you please check on unused imports ? Also in other changed classe (ex:- FlagdProviderTest.java )

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/FlagdProvider.java [hunk 3]
```diff
@@ -183,5 +190,12 @@ private void handleStateTransition(ProviderState oldState, ProviderState newStat
             this.emitProviderReady(details);
             this.emitProviderConfigurationChanged(details);
         }
+        // we shutdown from an error
+        if (ProviderState.ERROR.equals(oldState) && ProviderState.NOT_READY.equals(newState)) {
```
Comment (Kavindu-Dodan) [inline providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/FlagdProvider.java]: This condition will not be reached as there's `if (ProviderState.NOT_READY.equals(newState))` in line 167

Comment (Kavindu-Dodan) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/FlagdProvider.java]: Besides, this check is not needed. The state can transit from Error to Shutdown [1]

[1] - https://openfeature.dev/specification/sections/flag-evaluation#17-provider-lifecycle-management

File: providers/flagd/src/test/java/dev/openfeature/contrib/providers/flagd/FlagdProviderTest.java (hunk 1 above)
Comment (mehtasankets) [reply providers/flagd/src/test/java/dev/openfeature/contrib/providers/flagd/FlagdProviderTest.java]: Fixed.

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/FlagdProvider.java (hunk 2 above)
Comment (mehtasankets) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/FlagdProvider.java]: Fixed.

File: providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/FlagdProvider.java (hunk 3 above)
Comment (mehtasankets) [reply providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/FlagdProvider.java]: Fixed.

**Review by Kavindu-Dodan — APPROVED** [review]
Nice work @mehtasankets

File: mvnw
Comment (Kavindu-Dodan) [inline mvnw:1]: Not sure why there's a change here. @mehtasankets could you please check ?

Comment (mehtasankets) [reply mvnw:1]: I just made it executable. i.e. ran `chmod +x mvnw`.

Comment (Kavindu-Dodan) [reply mvnw:1]: Ohh, noted 👍

**Review by toddbaert — APPROVED** [review]
Thanks!

**Pull request merged**